### PR TITLE
Heatsink

### PR DESCRIPTION
**Pull request by Ferociousfeind**
**Feature:** This PR implements a heat-sink attribute.

## Feature Details
`MaximumHeat()` is used throughout the game, for all sorts of calculations. This makes adjusting it extremely easy. If an outfit has `"heatsink" 1`, then the ship it is installed on will have +100 maximum heat capacity (mass does the same thing, while also slowing down thrusters)

## UI Screenshots
N/A

## Usage Examples
```
outfit "Large Heat Shunt"
	category "Systems"
	cost 790000
	thumbnail "outfit/large heat shunt"
	"mass" 24
	"outfit space" -24
	"cooling" 54
	"heatsink" 36
	description "Without these powerful cooling systems, most Korath warships would overheat within seconds as soon as they began to fire their weapons."
```

## Testing Done
Putting a large amount of "heatsink" onto the cooling ducts outfit, and then putting a large number of those onto a ship. This vastly increases the ship's thermal capacity, and firing weapons with huge firing heat causes a much smaller jump on the right-hand-side heat bar than if there were no cooling ducts on the ship. The attribute works as intended.

## Performance Impact
I do not know if accessing attributes is expensive (it is probably not), most likely this does not impact performance.

### CODE REVIEW

**Comment by petervdmeer** [comment]
I don't think `"heatsink"` is a proper name for this attribute, because a heatsink is a device that transfers heat away from some fixed location to a flowing medium.
This attribute acts more like `"heat storage capacity"` (or `"additional heat storage capacity"`). The best name I can come up with is `"heat capacity"`, but that might be a bit confusing as well, since capacity is quite a common word. Please think of a better name for this attribute.

And what is the rationale behind adding a factor 100 difference between the value and the effect?

**Comment by Amazinite** [comment]
> And what is the rationale behind adding a factor 100 difference between the value and the effect?

Because that's what mass does. For each additional ton of mass that a ship has, the ship's maximum heat goes up by 100. This new attribute allows you to alter a ship's maximum heat the same way that mass does without needing to also effect everything else that mass effects. How the player will see this affect them is via the heat capacity row/column on their ship's energy/heat table, so I think that `"heat capacity"` is a fine attribute name. This also makes it easily analogous to `"energy capacity"`.
image

File: data/tooltips.txt [hunk 1]
```diff
@@ -221,6 +221,9 @@ tip "heat generation:"
 tip "heat protection:"
 	`Protection against weapons that do heat damage. If you add more than one outfit with this attribute, each additional one is less effective: a total value of 1 results in a 1 percent reduction in heat damage, while a total value of 11 only results in a 10 percent reduction.`
 
+tip "heatsink:"
+	`Materials with high specific heat can make for powerful heatsinks. One "heatsink" is equal to one ton of mass, for holding heat.`
```
Comment (Amazinite) [inline data/tooltips.txt:225]: The mass tooltip should probably also mention mass' influence on heat capacity.
```suggestion
tip "heat capacity:"
	`Increases the maximum heat that this ship is able to withstand before becoming overheated. A value of one has the same influence on a ship's maximum heat as one ton of mass, without the effects that mass has on movement.`
```

File: source/Ship.cpp [hunk 2]
```diff
@@ -3197,7 +3197,7 @@ double Ship::HeatDissipation() const
 // Get the maximum heat level, in heat units (not temperature).
 double Ship::MaximumHeat() const
 {
-	return MAXIMUM_TEMPERATURE * (cargo.Used() + attributes.Mass());
+	return MAXIMUM_TEMPERATURE * (cargo.Used() + attributes.Mass() + attributes.Get("heatsink"));
```
Comment (Amazinite) [inline source/Ship.cpp:3200]:
```suggestion
	return MAXIMUM_TEMPERATURE * (cargo.Used() + attributes.Mass() + attributes.Get("heat capacity"));
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

**Comment by Ferociousfeind** [comment]
If it's called "heat capacity", I would push a bit more for it to appear more analogous to energy capacity. If we compare energy capacity to energy generation, and then heat capacity to heat generation, and the effects they have on each other, then in-game "heat capacity" should appear 100x the raw value. There's a list of special displays for outfit attributes (60x, 60x60x, 60x100x, 100x + "%"...) is there are bare 100x variant?

**Comment by Amazinite** [comment]
`{"heat capacity", 4},`
Remember though that the ship's heat dissipation also affects the max heat a ship can withstand, so an outfit with a heat capacity of 1 (showing in game as 100) wouldn't actually increase your ship's max heat by 100.
The tooltip saying that a value of one heat capacity = one mass' worth of max heat would also need changed if we had heat capacity display as 100x.

**Comment by Amazinite** [comment]
Adding 1 ton to a Sparrow is +4.8 max heat. Adding 1 ton to a Bactrian is +2.4. So multiplying the displayed heat capacity be 100 would appear even less directly analogous.

**Comment by Ferociousfeind** [comment]
Okay. I'll refrain from that. Just your suggested changes should work, then.

**Review by Amazinite — APPROVED** [review]

**Pull request merged**